pika/connection.py

```diff
@@ -1102,14 +1149,15 @@ def _deliver_frame_to_channel(self, value):
 
         """
         if not value.channel_number in self._channels:
-            if self._is_basic_deliver_frame(value):
-                self._reject_out_of_band_delivery(value.channel_number,
-                                                  value.method.delivery_tag)
-            else:
-                LOGGER.warning("Received %r for non-existing channel %i", value,
-                               value.channel_number)
+            # This should never happen and would constitute breach of the
+            # protocol
+            LOGGER.critical(
+                'Received %s frame for unregistered channel %i on %s',
+                value.NAME, value.channel_number, self)
```

Comment:
@gmr, regarding
> Hmm I seem to remember something about consuming with noack=True with messages in the socket recv buffer prior to the Channel.Close being issued. Even with a high QoS prefetch, could you not have frames that have yet to be decoded when you go to close a channel?

This would indicate protocol error either in client or in broker. Here is why: `value.channel_number in self._channels` should continue to evaluate to TRUE until the channel is fully closed: `Channel.Close` to broker and then matching `Chanel.CloseOk` from broker or an unsolicited `Channel.Close` from broker. In either case, there should be no more messages destined to the channel after the channel receives either  `Channel.CloseOk` or `Channel.Close` from the broker, because the `Channel.CloseOk` or `Channel.Close` from the broker effectively flush the channel.

It turns out that there used to be a bug in pika that caused pika to remove a Channel prematurely in one condition, allowing such a race condition as you described to occur. I fixed that bug in this PR. The bug was in `Connection._close_channels`. There is a REVIEWME-tagged comment in this PR that describes the problem and the fix. Duplicating it here for completeness:

Regarding fix in `Connection._close_channels`:
> Now also initiates closing of channels that are in OPENING state (in addition to those in OPEN state). Also doesn't force-close channels that are in CLOSING state, allowing them to finish closing.
The legacy behavior lead to a race condition whereby frames could arrive on a channel that was prematurely removed by this function.